### PR TITLE
qownnotes: add tests

### PR DESCRIPTION
**Pull request by pbek**
###### Description of changes

This adds CLI and UI tests for QOwnNotes.
Loosely based on https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/keepassxc.nix (thank you for the insight!).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/tests/qownnotes.nix [hunk 1]
```diff
@@ -0,0 +1,72 @@
+import ./make-test-python.nix ({ pkgs, ...} :
+
+{
+  name = "qownnotes";
+  meta = with pkgs.lib.maintainers; {
+    maintainers = [ pbek ];
+    timeout = 1800;
```
[the remark this answers is not in the document]
Comment (pbek) [reply nixos/tests/qownnotes.nix]: Yes, it is. Thank you!

File: nixos/tests/qownnotes.nix [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+  nodes.machine = { ... }:
+
+  {
+    imports = [
+      ./common/user-account.nix
+      ./common/x11.nix
+    ];
+
+    services.xserver.enable = true;
```
[the remark this answers is not in the document]
Comment (pbek) [reply nixos/tests/qownnotes.nix]: Thank you!

File: nixos/tests/qownnotes.nix [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    qt = {
+      enable = true;
+      platformTheme = "kde";
+      style = "breeze";
+    };
```
[the remark this answers is not in the document]
Comment (pbek) [reply nixos/tests/qownnotes.nix:21]: No, thank you!

File: nixos/tests/qownnotes.nix [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        machine.wait_for_text("Panel layout")
+        machine.send_key("ret")
+        machine.wait_for_text("Nextcloud")
+        machine.send_key("ret")
+        machine.wait_for_text("App metric")
+        machine.send_key("ret")
+
+        # The main window should now show up
+        machine.wait_for_text("QOwnNotes - ${pkgs.qownnotes.version}")
+        machine.wait_for_console_text("QOwnNotes server listening on port 22222")
```
[the remark this answers is not in the document]
Comment (pbek) [reply nixos/tests/qownnotes.nix:65]: Great idea, thank you!

**Comment by NickCao** [comment]
> @NickCao, thanks a lot for your feedback! All pushed now. Any more suggestions?

Nothing particular from me, let's see if others have any ideas.

**Comment by pbek** [comment]
Result of `nixpkgs-review pr 237933` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

File: nixos/tests/qownnotes.nix [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        machine.wait_for_x()
+
+    with subtest("Check QOwnNotes version on CLI"):
+        ${aliceDo "qownnotes --version >&2 &"}
+
+        machine.wait_for_console_text("QOwnNotes ${pkgs.qownnotes.version}")
+
+    with subtest("Ensure QOwnNotes starts"):
+        # start QOwnNotes window
+        ${aliceDo "qownnotes >&2 &"}
```
Comment (illustris) [inline nixos/tests/qownnotes.nix:40]:
```suggestion
        ${aliceDo "qownnotes"}
```
same as above

**Comment by pbek** [comment]
> Result of nixpkgs-review pr 237933 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

I wonder if nixpkgs-review built the tests, the package or both...

**Review by illustris — APPROVED** [review]
One minor suggestion, looks good otherwise.

**Comment by NickCao** [comment]
@ofborg test qownnotes

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: nixos/tests/qownnotes.nix [hunk 6]
```diff
@@ -0,0 +1,72 @@
+import ./make-test-python.nix ({ pkgs, ...} :
```
[the remark this answers is not in the document]
Comment (pbek) [reply nixos/tests/qownnotes.nix:1]: Thank you very much!

**Comment by pbek** [comment]
Thank you for your feedback, @AndersonTorres!

**Comment by wineee** [comment]
Set `passthru.tests`  can help us automatically trigger this test when upgrading the package
An example is: https://github.com/NixOS/nixpkgs/blob/4c8ca604aef8204145c185c89cc52ee54dd7fc1a/pkgs/applications/window-managers/cage/default.nix#LL43C2-L43C18

**Review by wineee — APPROVED** [review]

**Comment by pbek** [comment]
@wineee, thank you! I'll add that with the `qownnotes: 23.6.4 -> 23.6.5` update. 😉

**Comment by NickCao** [comment]
@ofborg test qownnotes

**Comment by pbek** [comment]
Thank you all for your valuable insight!

**Pull request merged**